external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java

```diff
@@ -0,0 +1,114 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.storm.kafka;
+
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.storm.kafka.spout.internal.OffsetManager;
+import org.apache.storm.metric.api.IMetric;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+public class KafkaUtils {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaUtils.class);
+
+    public static class KafkaOffsetMetric implements IMetric {
+        Map<TopicPartition, OffsetManager> offsetManagers;
+        KafkaConsumer kafkaConsumer;
+
+        public KafkaOffsetMetric(Map<TopicPartition, OffsetManager> offsetManagers, KafkaConsumer kafkaConsumer) {
+            this.offsetManagers = offsetManagers;
+            this.kafkaConsumer = kafkaConsumer;
+        }
+
+        private class TopicMetrics {
+            long totalSpoutLag = 0;
+            long totalEarliestTimeOffset = 0;
+            long totalLatestTimeOffset = 0;
+            long totalLatestEmittedOffset = 0;
+            long totalLatestCompletedOffset = 0;
+        }
+
+        @Override
+        public Object getValueAndReset() {
```

Comment:
Remind me if this method is called by a different thread from the executor thread that runs nextTuple and the other spout methods? If so we need to add some synchronization on the offset managers. If not I don't think we need a ConcurrentHashMap in the spout.

Reply:
Thats good point. Will check  if it is called from same thread or different thread, based on that I will remove ConcurrentHashMap usage. We also don't need separate consumer for the metric, it it is called from the same thread. Will check and update the PR.

Reply:
@omkreddy , It seems both are invoked from the spout executor thread. The getValueAndReset is invoked from the tuple-action-fn which is invoked from [here](https://github.com/apache/storm/blob/1.x-branch/storm-core/src/clj/org/apache/storm/daemon/executor.clj#L627) and the nextTuple [here](https://github.com/apache/storm/blob/1.x-branch/storm-core/src/clj/org/apache/storm/daemon/executor.clj#L647)